### DISCUSSION TITLE
Set node value for polling without communicating it to the subscription.

### DISCUSSION THREAD

**Discussion by nikiFoe** [FreeOpcUa/python-opcua#1195] [Q&A]
Hello,

I am currently converting a rest-API in xml-form to an opcua-server. In my fist version I only used subscriptions when I wanted to get a node value. Now I need a mixture of polling and subscribing. The problem, I want to set the node value for polling every 30 seconds and I dont want subscibed nodes to notice this change . When I subscribe a node from the client I take the sampling interval and set the value accordingly. Because the 30 seconds from polling and the sampling intervall are not fitting together the nodes from the subscription are set even when they are not supposed to.

Is there a way to set subscribed nodes seperated from the polling process?

**Comment by AndreasHeine** [comment]
so you want do write a servers nodevalue from clientside (or serverside?) and the change of the value should not trigger a datachange notification?


serverside writing without a notification is possible!
```python
from opcua import Server, ua
from time import sleep

server = Server()

var = server.get_objects_node().add_variable(0, "Test", 0.0, ua.VariantType.Float)

dv = ua.DataValue(ua.Variant(0.0, ua.VariantType.Float))

server.start()

count = 0.0

while 1:
    sleep(1)
    count += 0.1
    #dv = ua.DataValue(ua.Variant(count, ua.VariantType.Float)) # this would trigger a datachange event!
    dv.Value = ua.Variant(count, ua.VariantType.Float) #this does not trigger because the classinstance is still the same!
    var.set_value(dv)
```

**Reply by nikiFoe** [reply]
I want to set the node value only on serverside. Yes, at least the value change in the "polling-loop" should not Influence the subscribed nodes .

**Reply by AndreasHeine** [reply]
like example above!

https://github.com/FreeOpcUa/python-opcua/blob/1cca33dd5fb70858b5e811b08356ca97520bcbaf/opcua/common/node.py#L195

```
        WARNING: On server side, ref to object is directly saved in our UA db, if this is a mutable object
        and you modfy it afterward, then the object in db will be modified without any
        data change event generated

```

**Reply by nikiFoe** [reply]
Thank, that works perfectly.

**Reply by nikiFoe** [reply]
I have to add a question. How is polling and monitoring possible at the same time? When setting the value of a node without triggering a datachange notification (polling-loop), the node has already set to the value, which means even when setting the same node to trigger a datachange notification with the same value it has already set to in the polling loop, I don't get a notification.

**Comment by nikiFoe** [comment]
I have a follow-up question. Can I set the timestamps of the node somehow?

When using the version which triggers a data change, I set the timestamp like this:

```
a = ua.uatypes.DataValue()
a.SourceTimestamp = xmlStatus[1]
a.ServerTimestamp = serverTimestamp
```

when I want to suppress the datachange-notification, I don't know how to set my own timestamp.

**Comment by AndreasHeine** [comment]
https://reference.opcfoundation.org/v104/Core/docs/Part4/7.7.4/
I hope you stick to the opc ua specification with what you are trieng to do!
SourceTimestamp ->` This is in contrast to the sourceTimestamp in which only the originator of the data is allowed to apply the sourceTimestamp.`

https://github.com/FreeOpcUa/python-opcua/blob/1cca33dd5fb70858b5e811b08356ca97520bcbaf/opcua/common/node.py#L195

the only way to change timestamps without a notification is if you pass the same classinstance of `ua.DataValue` to the` node.set_value` method. 

what is the actual question do you get an error?

**Reply by nikiFoe** [reply]
Yes, I try to stick with the OPC UA specification. But in my program, I get multiple XML-files with the values for the nodes. But I don't instantly update the node values, that happens just every 5 seconds, while the XML's are received at different intervals. I wanted to set the server timestamp to the time I received the nodes.

**Comment by starturtle** [comment]
I start getting really confused about when you do want change events and when you don't. Can you describe the full picture? It seems a little inconsistent overall, or maybe I'm not getting the pieces together properly.
Actually, I'd expect change events on timestamp update if subscribed with a trigger that includes them. If a client only cares about the value, the requested trigger type should be value-only. This should not be an issue to solve server-side. In particular if you do update the timestamp even when the value doesn't change, this can be an option to allow notification on change or notification on fresher values. 
You can, however, set the minimal publishing interval to your update rate, then it's also transparent to your users that they won't get data more often than that.

**Reply by nikiFoe** [reply]
Ok maybe I should explain the whole cocept more detailed.
I try to convert an rest API which provides multiple xmls, in which are always multiple values I want to set to nodes. On my server I have a loop that requests the xmls every 5 seconds and saves a timestamp to every single response.
I ask about the `set_value` without trigger a data change because I want to have a second loop that get specific values from the files and set them to the belonging node for polling every 30 seconds as well. My thrird loop is supposes to start when a client subscribed to a/multiple nodes. This monitoring loop does take the requests sampling intervall from the subscription and updates the node accordingly. I would be good, if the server can handle subscription and polling simultaniously. Because the Monitoring loop and the polling loop use the same nodes I need to set a node value without data trigger (the polling loop should not trigger a data change event). And also because the time I set the values  of the node are not always the time I received the xmls I want to set the timestamp manuelly to the time I received the xmls.

**Reply by starturtle** [reply]
Is there a reason why you don't limit it to evaluating changes made by the other two loops on the third, or adjust the minimal publishing interval your server is willing to accept, in case that's implemented according to your needs? That sounds like a behavior you're trying to achieve: use the values from the polling loops, including time stamps, but don't trigger events outside the subscription's sampling (publishing?) interval?
Sorry if neither is viable, I'm only familiar with the client side of this SDK.
From what you describe it's not fully clear to me what qualifies as "a data change" on the subscription (in particular, if polling updates the DataValue's content in the middle of a PublishingInterval, only very many things that aren't supposed to immediately trigger a data change event.
If automating this is not possible, does that effectively mean you wish to maintain two Values per Variable, one to return on read, and one to evaluate for MonitoredItems? It sounds a bit overly complicated, in particular if you have internal, unconditional polling already anyway. But having two Values logically is what it sounds like to me right now. Otherwise you'll always have the value to Read overwritten by whichever of the invocations comes last (triggering subscription evaluation or not). If that doesn't matter, it shouldn't be a problem.

**Comment by AndreasHeine** [comment]
@starturtle the timestamp has nothing to do with the trigger of a notification its just an extra information about the age of the data. a notification is only triggert if the datavalue class instance changes if only a class prop change no notification is triggerd

**Reply by starturtle** [reply]
That is the implementation side of when a server evaluates whether to send change events. It is an information of Python behavior.

What I'm talking about is the DataChangeTrigger member of the DataChangeFilter applied to MonitoredItems (See spec part 4, sections 5.12.1.4 and 7.17.2), which could probably simplify the whole matter a lot server-side (by transferring the responsibility to the client and evaluating each time), in particular if the triple update loop described above can end up with false positives and false negatives otherwise.
